TDDByExample/money/src/Money.js

```diff
@@ -0,0 +1,121 @@
+export class Expression {
+  reduce(bank, to) {
+    throw new Error(`reduce must be implemented ${bank} ${to}`);
+  }
+
+  plus(addend) {
+    throw new Error(`plus must be implemented ${addend}`);
+  }
+
+  times(multiplier) {
+    throw new Error(`times must be implemented ${multiplier}`);
+  }
+}
+
+export class Money extends Expression {
+  #amount = 0;
+
+  #currency = '';
+
+  constructor(amount, currency) {
+    super();
+    this.#amount = amount;
+    this.#currency = currency;
+  }
+
+  get amount() {
+    return this.#amount;
+  }
+
+  times(multiplier) {
+    return new Money(this.amount * multiplier, this.#currency);
+  }
+
+  get currency() {
+    return this.#currency;
+  }
+
+  equals(another) {
+    if (!(another instanceof Money)) {
+      return false;
+    }
+    if (this.currency !== another.currency) {
+      return false;
+    }
+
+    return this.amount === another.amount; // 1. private이기 때문에 에러가 발생한다.
+  }
+
+  toString() {
+    return `${this.amount} ${this.currency}`;
+  }
+
+  reduce(bank, to) {
+    const rate = bank.rate(this.currency, to);
+
+    return new Money(this.amount / rate, to);
+  }
+}
+
+export class Sum extends Expression {
+  augend = null;
+
+  addend = null;
+
+  constructor(augend, addend) {
+    super();
+    this.augend = augend;
+    this.addend = addend;
+  }
+
+  reduce(bank, to) {
+    const amount = this.augend.reduce(bank, to).amount + this.addend.reduce(bank, to).amount;
+
+    return new Money(amount, to);
+  }
+
+  plus(addend) {
+    return new Sum(this, addend);
+  }
+
+  times(multiplier) {
+    return new Sum(this.augend.times(multiplier), this.addend.times(multiplier));
+  }
+}
+
+Money.dollar = function(amount) {
+  return new Money(amount, 'USD');
+};
+
+Money.franc = function(amount) {
+  return new Money(amount, 'CHF');
+};
+
+Money.prototype.plus = function plus(addend) {
+  if (this.currency === addend.currency) {
+    return new Money(this.amount + addend.amount, this.currency);
+  }
+
+  return new Sum(this, addend);
+};
+
+export class Bank {
+  #rates = {};
+
+  reduce(source, to) {
+    return source.reduce(this, to);
+  }
+
+  addRate(from, to, rate) {
+    this.#rates[`${from} ${to}`] = rate;
```

Comment:
이렇게 from-to 통화 이름을 객체의 key로 사용하는건 신박하네요 ㅎㅎ
하지만 저라면 조금 다르게 자료 구조를 잡을 것 같아요.
최대한 논리적으로 이해하기 쉽고, 코드에 의도가 드러나도록 아래처럼 사용할 것 같습니다.

```suggestion
    // ... 대충 `this.#rates`에 `from` 값이 없으면 `{ratio: {}}`객체를 생성해서 할당하는 코드
    this.#rates[from].ratio[to] = rate;
```
결국 특정 통화를 다른 통화로 환산할 때 얼만큼의 상대적 비율을가지는지만 저장하면 되니까요 ㅎㅎ